### ISSUE TITLE
hgpak compress

### ISSUE THREAD

**Issue by czdvoika** [monkeyman192/HGPAKtool#1]
hi there :) can you create python script "pack" without compress? like fonts tnx :)

**Comment by monkeyman192** [comment]
I think this should already be possible, but I think there was a bug that meant it always compressed, because default was True, and including the flag meant True also.
I already have the change, so I'll push the code and you can try it out.
Basically by default it won't compress and you'd need to add `--compress` for it to actually compress the data.

**Comment by czdvoika** [comment]
great thanks a lot, experimentation can begin :) im not a programmer or revers it is correct? 
[https://ibb.co/cyMsZRL](url)

**Comment by monkeyman192** [comment]
Have you tried the `--repack` option? It should repack in the original order

**Comment by czdvoika** [comment]
this time I used it and you probably have a bug, everything fits and is the same except the hash, see photo.
https://ibb.co/DLL6jCJ

**Comment by czdvoika** [comment]
I used it this time, don't you have a bug? I repackaged the fonts without any changes and the files are the same except for the hash.
https://ibb.co/gjm7cdy

**Comment by monkeyman192** [comment]
ah yeah, I think the hash will be different because I don't know what hashing algorithm is used.
Keep in mind that this tool isn't actually tested on the switch files as I can't mod my switch (nor do I have NMS for it...)
This repo was mostly just to explore the possibility so that if it were possible it would help. It may just not be possible...

**Comment by czdvoika** [comment]
Thanks for your time, I'll write again :)

**Comment by czdvoika** [comment]
i get it:) md5 file path
https://ibb.co/JQRfp2P

**Comment by czdvoika** [comment]
the game doesn't freeze this time but there's still a bug somewhere...
https://ibb.co/s90FQ9S

**Comment by monkeyman192** [comment]
Brilliant! Let me implement that today! Thanks so much!
Unless you'd like to create a PR for it?

**Comment by czdvoika** [comment]
I don't need anything, just make it work :D

**Comment by monkeyman192** [comment]
I'm a little confused actually... I am looking at the vanilla .pak files, and the hashes don't seem to match what I'd expect if I used md5 to hash them.
Can you send a vanilla pak file for me to look at? The only switch ones I have are the original release, so maybe they changed it at some point?

**Comment by czdvoika** [comment]
here we go https://drive.google.com/file/d/1RRc-dugGBE8NVB00V8zsifmZU__fV4l2/view?usp=sharing
I tried it again and it fits.
https://ibb.co/R90dY1S

**Comment by monkeyman192** [comment]
OK, so I can generate the hashes fine for each of the files, my only issue right now is that the first "file" is the list of names and I can't figure out how to generate the Hash for this. I looked at the binary and it seems like it might be hashing the name of the file, but I tried  bunch of different combos and nothing worked... it's also worth noting that the switch and macs have the same Hash for this "file"

**Comment by czdvoika** [comment]
I don't know what the first hash is supposed to mean either, but we're close :) tnx bro you badboy

**Comment by monkeyman192** [comment]
Yeah, neither... My solution in the end was to just extract them all and have a lookup!
It will work for repacking vanilla files, but won't work for creating new modded paks unfortunately...

**Comment by monkeyman192** [comment]
I have just pushed a change. Please let me know how it goes.

**Comment by czdvoika** [comment]
wou repack identical, good job:)

**Comment by czdvoika** [comment]
hehe it works, I already have the game in my native language(czech) :) tnx a lot
https://ibb.co/cgR9ptn

**Comment by nucleusxx** [comment]
hi help me pls how to pack the font file back


fonts is not a valid path to extract.
Unpacked 0 files from 0 .pak's in 0.000206s

**Comment by monkeyman192** [comment]
If you are talking about on a console, can you provide the BankSignatures.bin file you have?
On pc at least the file is encrypted (but we can get around this...)
I just decrypted the latest one for steam and I can see that it's not actually signatures or any kind of hash but it is in fact just the size of the .bnk file in bytes.
If this is the same on console, then it would be easy enough to modify the values and re-encrypt it

**Comment by Zigler** [comment]
I think based upon the zstd specification there should be a way to determine parameters for that. I also know that a block appears to be padded out to about  0x10000 bytes. Since the frame header has quite a bit of information in it already there is a way to set up zstandard to at least operate with those settings for the windows pak files.

**Comment by monkeyman192** [comment]
I did have a good look through that doc and tried to parse the block header to determine what was used, but I couldn't figure it out...
If you figure anything out that would be great.
I have almost completed the work on adding the repacking, so will have a PR up for it soon that you can try out.

**Comment by monkeyman192** [comment]
Hmm interesting. I am not exactly sure how to read the data in that text file.
One thing the bad magic could be is actually uncompressed chunks - occasionally HG will have uncompressed chunks strewn in the middle of compressed ones... No idea why but it can happen even if the pak has the flag to mark itself as "compressed"

**Comment by Zigler** [comment]
> Hmm interesting. I am not exactly sure how to read the data in that text file. One thing the bad magic could be is actually uncompressed chunks - occasionally HG will have uncompressed chunks strewn in the middle of compressed ones... No idea why but it can happen even if the pak has the flag to mark itself as "compressed"

So I had structured it as the zstd c library calls the function. The main thing to look at there is the compressionLevel. Here's an excerpt of the c++ I wrote to figure out the settings. I pretty much figured based on everything else I had seen that it wasn't using a dictionary file or custom window parameters. Excuse the mess of code.
```
    size_t const dSize = ZSTD_decompress(rBuff, rSize, cBuff, cSize);
    CHECK_ZSTD(dSize);
    CHECK(dSize == rSize, "Impossible because zstd will check this condition!");
    size_t const ccBuffSize = ZSTD_compressBound(dSize);
    void* const ccBuff = malloc_orDie(ccBuffSize);
    uint8_t* caBuf  = reinterpret_cast<uint8_t *>(ccBuff);
    uint8_t* cbBuf  = reinterpret_cast<uint8_t *>(cBuff);
    for(int i=1;i<23;i++){
        size_t const ccSize = ZSTD_compress(ccBuff, ccBuffSize, rBuff, dSize, i);
        if(ccSize==cSize){
            bool isCorrect = true;
            for(int j=0;j<ccSize;j++){
                if(caBuf[j]!=cbBuf[j]){
                    isCorrect = false;
                    break;
                }
            }
            if(isCorrect){
                printf("\nZSTD_compress(dst=(void *)0x%016lX,dstCapacity=%i,src=(void *)0x%016lX,srcSize=%i,compressionLevel=%i) Size: %" PRIu64,ccBuff, ccBuffSize, rBuff, dSize, i,ccSize);
            }
        }
        CHECK_ZSTD(ccSize);
    }
    printf("\n");
```

**Comment by monkeyman192** [comment]
ok thanks for this!
I realised the issue... Because I was testing this on windows when I wrote `\r\n` to the file as a string it would expand the `\n` to `\r\r`, resulting in a manifest with `\r\r\n` as the "newline character".
This would mean the entire compression would differ.
Once I changed it to write the manifest in bytes explicitly this issue went away and now it recompresses perfectly!
I'm not sure if I'll have time to push the changes tonight as I want to write some extra tests for this, but I'll update the repo tomorrow with this fix, thanks!
